// @arshaw/monorepo-tool/dist/src/pkg/parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function parseAllPkgInput(allPkgInput, rootConfig) {
    let allPkgInputArray = Array.isArray(allPkgInput)
        ? allPkgInput
        : (allPkgInput ? [allPkgInput] : []);
    return allPkgInputArray.map((pkgInput) => combinePkgConfigs(rootConfig, parsePkgInput(pkgInput)));
}
exports.parseAllPkgInput = parseAllPkgInput;
function parseConfig(input) {
    let distDir = input.distDir || '';
    let ignoreFiles = Array.isArray(input.ignoreFiles)
        ? input.ignoreFiles
        : (input.ignoreFiles ? [input.ignoreFiles] : []);
    let npmClientArgs = input.npmClientArgs || [];
    return { distDir, ignoreFiles, npmClientArgs };
}
exports.parseConfig = parseConfig;
function parsePkgInput(input) {
    if (typeof input === 'string') {
        return Object.assign(parseConfig({}), { path: input });
    }
    else {
        return Object.assign(parseConfig(input), { path: input.path || '' });
    }
}
exports.parsePkgInput = parsePkgInput;
function combinePkgConfigs(config0, config1) {
    return {
        path: config1.path,
        distDir: config1.distDir || config0.distDir,
        ignoreFiles: config0.ignoreFiles.concat(config1.ignoreFiles),
        npmClientArgs: config0.npmClientArgs.concat(config1.npmClientArgs)
    };
}
